File: src/jvmMain/java/ess/papyrus/ParamID.java
package ess.papyrus;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.nio.ByteBuffer;
import java.util.Objects;

/**
 * An opcode parameter that stores an identifier.
 */
final public class ParamID extends Parameter {

    public ParamID(TString val) {
        this.VALUE = Objects.requireNonNull(val);
    }

    @NotNull
    @Override
    public ParamType getType() {
        return ParamType.IDENTIFIER;
    }

    @Override
    public void write(@NotNull ByteBuffer output) {
        this.getType().write(output);
        this.VALUE.write(output);
    }

    @Override
    public int calculateSize() {
        return 1 + this.VALUE.calculateSize();
    }

    /**
     * @return String representation.
     */
    @NotNull
    @Override
    public String toValueString() {
        return this.VALUE.toString();
    }

    @Override
    public int hashCode() {
        int hash = 7;
        hash = 41 * hash + Objects.hashCode(this.getType());
        hash = 41 * hash + Objects.hashCode(this.VALUE);
        return hash;
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        if (this == obj) {
            return true;
        } else if (obj == null) {
            return false;
        } else if (getClass() != obj.getClass()) {
            return false;
        }
        final ParamID other = (ParamID) obj;
        return this.VALUE.equals(other.VALUE);
    }

    @Override
    public boolean isTemp() {
        return TEMP_PATTERN.test(this.VALUE.toString())
                && !AUTOVAR_PATTERN.test(this.VALUE.toString())
                && !NONE_PATTERN.test(this.VALUE.toString());
    }

    /**
     * @return A flag indicating if the parameter is an Autovariable.
     */
    @Override
    public boolean isAutovar() {
        return AUTOVAR_PATTERN.test(this.VALUE.toString());
    }

    /**
     * @return A flag indicating if the parameter is an None variable.
     */
    @Override
    public boolean isNonevar() {
        return NONE_PATTERN.test(this.VALUE.toString());
    }

    final public TString VALUE;
}

File: src/jvmMain/kotlin/ess/papyrus/ParamID.kt
package ess.papyrus

import java.nio.ByteBuffer


/**
 * An opcode parameter that stores an identifier.
 */
class ParamID(`val`: TString?) : Parameter() {
    override val type: ParamType
        get() = ParamType.IDENTIFIER

    override fun write(output: ByteBuffer?) {
        type.write(output)
        VALUE.write(output)
    }

    override fun calculateSize(): Int {
        return 1 + VALUE.calculateSize()
    }

    /**
     * @return String representation.
     */
    override fun toValueString(): String {
        return VALUE.toString()
    }

    override fun hashCode(): Int {
        var hash = 7
        hash = 41 * hash + type.hashCode()
        hash = 41 * hash + VALUE.hashCode()
        return hash
    }

    override fun equals(other: Any?): Boolean {
        return when {
            this === other -> {
                true
            }
            other == null -> {
                false
            }
            javaClass != other.javaClass -> {
                false
            }
            else -> {
                val other2 = other as ParamID
                VALUE.equals(other2.VALUE)
            }
        }
    }

    override val isTemp: Boolean
        get() = (TEMP_PATTERN.test(VALUE.toString())
                && !AUTOVAR_PATTERN.test(VALUE.toString())
                && !NONE_PATTERN.test(VALUE.toString()))

    /**
     * @return A flag indicating if the parameter is an Autovariable.
     */
    override val isAutovar: Boolean
        get() = AUTOVAR_PATTERN.test(VALUE.toString())

    /**
     * @return A flag indicating if the parameter is an None variable.
     */
    override val isNonevar: Boolean
        get() = NONE_PATTERN.test(VALUE.toString())
    val VALUE: TString = `val`!!

}